File: data/readable_format/fortran/file1963.f90
! Define the subroutine sla_PERMUT
SUBROUTINE sla_PERMUT(N, ISTATE, IORDER, J)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: N
  INTEGER, INTENT(INOUT) :: ISTATE(N), IORDER(N)
  INTEGER, INTENT(OUT) :: J
  INTEGER :: I, IP1, ISLOT, ISKIP

  IF (N.LT.1) THEN
     J = -1
     RETURN
  ELSE
     J = 0
  END IF

  IF (ISTATE(1).LT.0) THEN
     ISTATE(1) = -1
     DO I=2,N
        ISTATE(I) = 0
     END DO
  END IF

  ISTATE(1) = ISTATE(1) + 1

  DO I=1,N
     IF (ISTATE(I).GE.I) THEN
        ISTATE(I) = 0
        IF (I.GE.N) THEN
           J = 1
        ELSE
           IP1 = I + 1
           ISTATE(IP1) = ISTATE(IP1) + 1
        END IF
     END IF
  END DO

  DO I=1,N
     IORDER(I) = 1
  END DO

  DO I=N,2,-1
     ISLOT = 0
     DO ISKIP=0,ISTATE(I)
        ISLOT = ISLOT + 1
        DO WHILE (IORDER(ISLOT).GT.1)
           ISLOT = ISLOT + 1
        END DO
     END DO
     IORDER(ISLOT) = I
  END DO
END SUBROUTINE sla_PERMUT

! Main program to test sla_PERMUT subroutine
PROGRAM test_sla_PERMUT
  IMPLICIT NONE
  INTEGER, PARAMETER :: N = 4
  INTEGER :: ISTATE(N), IORDER(N), J, i

  ! Initialize ISTATE with a specific test case, e.g., all -2
  ISTATE = -2
  CALL sla_PERMUT(N, ISTATE, IORDER, J)

  PRINT *, 'Test Case: J = ', J
  PRINT *, 'ISTATE =', ISTATE
  PRINT *, 'IORDER =', IORDER

  ! Include further test cases as needed
END PROGRAM test_sla_PERMUT

File: data/readable_format/cpp/file1963.cpp
#include <gtest/gtest.h>
#include <vector>

// Definition of sla_PERMUT directly in this file
void sla_PERMUT(int N, std::vector<int>& ISTATE, std::vector<int>& IORDER, int& J) {
    int I, IP1, ISLOT, ISKIP;

    if (N < 1) {
        J = -1;
        return;
    } else {
        J = 0;
    }

    if (ISTATE[0] < 0) {
        ISTATE[0] = -1;
        for (I = 1; I < N; ++I) {
            ISTATE[I] = 0;
        }
    }

    ISTATE[0] += 1;

    for (I = 0; I < N; ++I) {
        if (ISTATE[I] >= I + 1) {
            ISTATE[I] = 0;
            if (I >= N - 1) {
                J = 1;
            } else {
                IP1 = I + 1;
                ISTATE[IP1] += 1;
            }
        }
    }

    for (I = 0; I < N; ++I) {
        IORDER[I] = 1;
    }

    for (I = N - 1; I >= 1; --I) {
        ISLOT = 0;
        for (ISKIP = 0; ISKIP <= ISTATE[I]; ++ISKIP) {
            ISLOT += 1;
            while (IORDER[ISLOT] > 1) {
                ISLOT += 1;
            }
        }
        IORDER[ISLOT] = I + 1;
    }
}

// Unit Test using Google Test
TEST(SlaPermutTest, BasicTest) {
    int N = 4;
    std::vector<int> ISTATE(N, -2); // Initialize ISTATE to trigger initialization
    std::vector<int> IORDER(N);
    int J;

    sla_PERMUT(N, ISTATE, IORDER, J);
    
    // Example assertion, needs to be adjusted based on the expected outcome
    ASSERT_EQ(J, 0); // This is a placeholder, adjust the expected value accordingly
    // Add more assertions here to check the contents of ISTATE and IORDER
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}